[BUG] Bug with limited API support in trivial program involving float literal

### Describe the bug

Float literals in equality comparison operators in limited API mode trigger some unsupported code path. Oddly, the compile error comes extremely early, in the include of Python.h itself, which made me slightly suspect a potential error in CPython's own limited API support. However, other operators (e.g. inequality comparisons) do not trigger the error, and neither do integer comparisons, which makes me think that it's more likely that the cause is something Cython itself is doing to generate invalid code that somehow triggers this path. Unfortunately, I can't quite pin down what it is.

I'll note that I also see different failure modes depending on whether I compile the Cython module as C or C++. When compiling in C++, I see this:
```
/home/coder/.conda/envs/rapids/bin/x86_64-conda-linux-gnu-cc -DNDEBUG -fwrapv -O2 -Wall -fPIC -O2 -isystem /home/coder/.conda/envs/rapids/include -fPIC -O2 -isystem /home/coder/.conda/envs/rapids/include -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /home/coder/.conda/envs/rapids/include -I/home/coder/.conda/envs/rapids/targets/x86_64-linux/include -L/home/coder/.conda/envs/rapids/targets/x86_64-linux/lib -L/home/coder/.conda/envs/rapids/targets/x86_64-linux/lib/stubs -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem /home/coder/.conda/envs/rapids/include -I/home/coder/.conda/envs/rapids/targets/x86_64-linux/include -L/home/coder/.conda/envs/rapids/targets/x86_64-linux/lib -L/home/coder/.conda/envs/rapids/targets/x86_64-linux/lib/stubs -fPIC -DPy_LIMITED_API=0x030B0000 -DCYTHON_LIMITED_API -I/home/coder/.conda/envs/rapids/include/python3.11 -c test.cpp -o build/temp.linux-x86_64-cpython-311/test.o
In file included from /home/coder/.conda/envs/rapids/include/python3.11/Python.h:55,
                 from test.cpp:41:
/home/coder/.conda/envs/rapids/include/python3.11/floatobject.h: In function 'PyObject* __Pyx_PyFloat_EqObjC(PyObject*, PyObject*, double, int, int)':
/home/coder/.conda/envs/rapids/include/python3.11/floatobject.h:14:26: error: 'PyFloat_Type' has incomplete type
   14 | PyAPI_DATA(PyTypeObject) PyFloat_Type;
      |                          ^~~~~~~~~~~~
In file included from /home/coder/.conda/envs/rapids/include/python3.11/Python.h:42,
                 from test.cpp:41:
/home/coder/.conda/envs/rapids/include/python3.11/pytypedefs.h:20:16: note: forward declaration of 'PyTypeObject' {aka 'struct _typeobject'}
   20 | typedef struct _typeobject PyTypeObject;
```

When compiling in C, I instead see this:
```
/home/coder/.conda/envs/rapids/bin/x86_64-conda-linux-gnu-cc -DNDEBUG -fwrapv -O2 -Wall -fPIC -O2 -isystem /home/coder/.conda/envs/rapids/include -fPIC -O2 -isystem /home/coder/.conda/envs/rapids/include -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /home/coder/.conda/envs/rapids/include -I/home/coder/.conda/envs/rapids/targets/x86_64-linux/include -L/home/coder/.conda/envs/rapids/targets/x86_64-linux/lib -L/home/coder/.conda/envs/rapids/targets/x86_64-linux/lib/stubs -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem /home/coder/.conda/envs/rapids/include -I/home/coder/.conda/envs/rapids/targets/x86_64-linux/include -L/home/coder/.conda/envs/rapids/targets/x86_64-linux/lib -L/home/coder/.conda/envs/rapids/targets/x86_64-linux/lib/stubs -fPIC -DPy_LIMITED_API=0x030B0000 -DCYTHON_LIMITED_API -I/home/coder/.conda/envs/rapids/include/python3.11 -c test.c -o build/temp.linux-x86_64-cpython-311/test.o
test.c: In function '__Pyx_PyFloat_EqObjC':
test.c:6745:37: error: invalid use of incomplete typedef 'PyTypeObject' {aka 'struct _typeobject'}
 6745 |                         PyFloat_Type.tp_richcompare(op2, op1, Py_EQ));
```

### Code to reproduce the behaviour:

```cython
# test.pyx
x = 1
# x = 1.0  # This is also fine

x == 1.0  # This will raise an error regardless of which line is used above
# x < 1.0  # This will not raise an error
# x == 1  # This will not raise an error

# setup.py
from setuptools import Extension, setup
from Cython.Build import cythonize

extensions = [
    Extension(
        "test", ["test.pyx"],
        define_macros=[("Py_LIMITED_API", "0x030B0000"), ("CYTHON_LIMITED_API", None)],
        language="c++",  # This changes the error
    )
]

setup(
    name='test',
    ext_modules=cythonize(extensions),
)
```

### Expected behaviour

This code should compile

### OS

Linux

### Python version

3.11.8

### Cython version

Merge of #6061 and the limited API branch from #6082

### Additional context

There is also a separate limited API issue with `PyList_GET_ITEM` that I observe in the above MWE, but I've fixed that in #6061.

--- a/Cython/Utility/Optimize.c
+++ b/Cython/Utility/Optimize.c
@@ -1481,8 +1481,19 @@ static {{c_ret_type}} {{cfunc_name}}(PyObject *op1, PyObject *op2, double floatv
                 default:
         #endif
         {{if op in ('Eq', 'Ne')}}
+                    {
+                        PyObject *res =
+                    #if CYTHON_USE_TYPE_SLOTS || __PYX_LIMITED_VERSION_HEX > 0x030A0000
+                            // PyType_GetSlot only works on non-heap types from Python 3.10
+                            __Pyx_PyType_GetSlot((&PyFloat_Type), tp_richcompare, richcmpfunc)
+                    #else
+                            PyObject_RichCompare
+                    #endif
+                        ({{'op1, op2' if order == 'CObj' else 'op2, op1'}},
+                         Py_{{op.upper()}});
                     return {{'' if ret_type.is_pyobject else '__Pyx_PyObject_IsTrueAndDecref'}}(
-                        PyFloat_Type.tp_richcompare({{'op1, op2' if order == 'CObj' else 'op2, op1'}}, Py_{{op.upper()}}));
+                        res);
+                    }
         {{else}}
                     {{fval}} = PyLong_AsDouble({{pyval}});
                     if (unlikely({{fval}} == -1.0 && PyErr_Occurred())) return NULL;

--- a/tests/run/isolated_limited_api_tests.srctree
+++ b/tests/run/isolated_limited_api_tests.srctree
@@ -21,6 +21,8 @@ setup(
 
 ##################### run.py ##################################
 
+import decimal
+
 import limited
 
 limited.fib(11)
@@ -49,6 +51,10 @@ assert limited.cast_float(bytearray(b"3")) == 3.0
 assert limited.add_one(2) == 3
 assert limited.float_equals(1.5)
 assert not limited.float_equals(2)
+assert not limited.float_equals(None)
+assert limited.float_equals(decimal.Decimal("1.5"))
+
+
 
 ##################### limited.pyx #############################
 

--- a/tests/run/isolated_limited_api_tests.srctree
+++ b/tests/run/isolated_limited_api_tests.srctree
@@ -47,7 +47,8 @@ assert limited.cast_float("2.0") == 2.0
 assert limited.cast_float(bytearray(b"3")) == 3.0
 
 assert limited.add_one(2) == 3
-
+assert limited.float_equals(1.5)
+assert not limited.float_equals(2)
 
 ##################### limited.pyx #############################
 
@@ -93,6 +94,9 @@ def catches():
     
 def cast_float(object o):
     return float(o)
+
+def float_equals(o):
+    return o == 1.5
 
 class C:
     pass

